File: frontend/src/services/productService.js
import API from "./api";



export const getAllProducts = async () => {

    try {

        const response = await API.get("/products");

        return response.data;


    } catch(error) {

        throw error;

    }

};



export const getProductsByCategory = async (category) => {

    try {

        const response = await API.get(
            `/products/category/${category}`
        );

        return response.data;


    } catch(error) {

        throw error;

    }

};






export const getProductById = async (id) => {

    try {

        const response = await API.get(
            `/products/${id}`
        );

        return response.data;


    } catch(error) {

        throw error;

    }

};






export const addProduct = async (productData) => {

    try {

        const response = await API.post(
            "/products/add",
            productData
        );

        return response.data;


    } catch(error) {

        throw error;

    }

};







export const updateProduct = async (id, productData) => {

    try {


        const response = await API.put(

            `/products/update/${id}`,

            productData

        );


        return response.data;



    } catch(error) {


        throw error;


    }

};







export const deleteProduct = async (id) => {

    try {


        const response = await API.delete(

            `/products/delete/${id}`

        );


        return response.data;



    } catch(error) {


        throw error;


    }

};

File: backend/src/main/java/com/shopsphere/backend/service/ProductService.java
package com.shopsphere.backend.service;

import com.shopsphere.backend.entity.Product;
import com.shopsphere.backend.repository.ProductRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ProductService {

    @Autowired
    private ProductRepository repository;

    // Add Product
    public Product addProduct(Product product) {

        calculatePrice(product);

        return repository.save(product);

    }

    // Get All Products
    public List<Product> getAllProducts() {

        return repository.findAll();

    }

    // Get Product By Id
    public Product getProduct(Long id) {

        return repository.findById(id).orElse(null);

    }

    // Search Products
    public List<Product> searchProducts(String keyword) {

        return repository.findByNameContainingIgnoreCase(keyword);

    }

    // Category Products
    public List<Product> getCategoryProducts(String category) {

        return repository.findByCategory(category);

    }

    // Brand Products
    public List<Product> getBrandProducts(String brand) {

        return repository.findByBrand(brand);

    }

    // Update Product
    public Product updateProduct(Long id, Product product) {

        Product old = repository.findById(id).orElse(null);

        if (old == null) {

            return null;

        }

        product.setId(id);

        calculatePrice(product);

        return repository.save(product);

    }

    // Delete Product
    public void deleteProduct(Long id) {

        repository.deleteById(id);

    }

    // Auto Calculate Price
    private void calculatePrice(Product product) {

        if (product.getBrandPrice() != null &&
                product.getSellingPrice() != null &&
                product.getBrandPrice() > 0) {

            double discount =
                    ((product.getBrandPrice() -
                            product.getSellingPrice())
                            / product.getBrandPrice()) * 100;

            product.setDiscount(
                    Math.round(discount * 100.0) / 100.0
            );

        }

        if (product.getSellingPrice() != null &&
                product.getGst() != null) {

            double finalPrice =
                    product.getSellingPrice() +
                            ((product.getSellingPrice()
                                    * product.getGst()) / 100);

            product.setFinalPrice(
                    Math.round(finalPrice * 100.0) / 100.0
            );

        }

    }

}